Azure Pipelines on DevOps with certificate-based authentication and Graph v2.0 not working

#### Details of the scenario you tried and the problem that is occurring

Note: I know this is not a problem with M365Dsc itself, but that's what I'm using in the pipelines and I don't know where else to report the problem.

Currently Windows containers running on Azure Pipelines on DevOps have Graph v1.28 pre-installed, when installing the latest M365Dsc version v1.23.705.1 they get updated to v2.0 (Update-M365DSCDependencies) and whatever is outdated is uninstalled (Uninstall-M365DSCOutdatedDependencies).

Nevertheless calling simple cmdlets such as Get-MgUser or Get-MgGroup ends up in error, see below, I cannot test every single resource to check which ones are affected by this but my solution relies on calling Get-MgGroup to translate GroupIds to DisplayNames and vice-versa due to the way Assignments work on Intune resources (only by GroupId) which makes upgrading not an option, which at the same time makes me also not able to test the other issues I reported with the latest version.

I already tested this unsuccessfully:
	- Removing all Microsoft.Graph.* v1.28 from the session
	- Removing all their associated folders from "C:\Program Files\WindowsPowerShell\Modules"
	- Install Microsoft.Graph.Authentication v2.0.0
	- Install Microsoft.Graph.Groups v2.0.0
	- Install MSCloudLoginAssistant v1.0.116
	- Connect to Graph using Connect-M365Tenant with certificate
	- Calling Get-MgGroup, error happens here

#### Verbose logs showing the problem

```powershell
##[error]Method not found: 'Void Microsoft.Graph.Authentication.AzureIdentityAccessTokenProvider..ctor(Azure.Core.TokenCredential, System.String[], Microsoft.Kiota.Authentication.Azure.ObservabilityOptions, System.String[])'
```

#### Suggested solution to the issue

Update Windows container image to have Graph v2.0 by default instead of v1.28.0.

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)

N/A

#### The operating system the target node is running

OsName               : Microsoft Windows Server 2022 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 20348.1.amd64fre.fe_release.210507-1500
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

#### Version of the DSC module that was used ('dev' if using current dev branch)

1.23.705.1

Having the same issue here. My pipeline in Azure DevOps is executing a Powershell script that has Graph commands over a public DevOps agent and almost any Graph command ends up in this error message: ##[error]Method not found: 'Void Microsoft.Graph.Authentication.AzureIdentityAccessTokenProvider..ctor(Azure.Core.TokenCredential, System.String[], Microsoft.Kiota.Authentication.Azure.ObservabilityOptions, System.String[])'

I've done various tests, including the ones mentioned: what I can tell you for sure is that as long as Microsoft.Graph v2.0.0 or v1.28.0 is installed, most commands end up with the same error. I've even tried uninstalling the current version that is pushed on the agent (whichever that might be) and installing v1.19.0 - same behavior.
Given this last test, I'm inclined to think this isn't a module issue, but some update that went wrong on the DevOps agents. If I corroborate this with the fact that Graph commands work properly locally, it sounds even more feasible.

@andikrueger @peombwa do you guys have any additional insight on this?

There seems to be a bigger issue now...

I've run several more tests and Graph v2.0 now is always available, nevertheless just by calling Export-M365DSCConfiguration on Intune workload it fails with:

```powershell
##[error]The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: Function Export-TargetResource cannot be created because function capacity 4096 has been exceeded for this scope.
```

To "solve" this the value of $MaximumFunctionCount must be increased, I'm able to set it to 32768 (its max value) but for some reason the value gets lost when it starts the export and then I get that error message above which basically means that without solution I'm stuck with running M365Dsc v1.23.628.1, on Azure Pipelines, which is the last one to support Graph v1.0, this is not good and will have to report this to my management.

This can be replicated locally with the following:

```powershell
$MaximumFunctionCount = 4096
Get-Module -ListAvailable Microsoft.Graph.* | `
    Select-Object -ExpandProperty name | `
    Sort-Object -Unique | `
    %{ Import-Module -Force $_ }
```

@andikrueger Probably you should change the label of this issue to Bug instead of Enhancement.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,13 @@
 # Change log for Microsoft365DSC
 
+# UNRELEASED
+
+* DEPENDENCIES
+  * Updated all Microsoft.Graph modules to version 2.1.0.
+  FIXES [3454](https://github.com/microsoft/Microsoft365DSC/issues/2454)
+* MISC
+  * Added a more meaningful exception message to Update-M365DSCDependencies if the module is not installed or imported successfully
+
 # 1.23.712.1
 
 * IntuneDeviceConfigurationPolicyAndroidDeviceAdministrator

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,7 +4,6 @@
 
 * DEPENDENCIES
   * Updated all Microsoft.Graph modules to version 2.1.0.
-  FIXES [3454](https://github.com/microsoft/Microsoft365DSC/issues/2454)
 * MISC
   * Added a more meaningful exception message to Update-M365DSCDependencies if the module is not installed or imported successfully
 

--- a/Modules/Microsoft365DSC/Dependencies/Manifest.psd1
+++ b/Modules/Microsoft365DSC/Dependencies/Manifest.psd1
@@ -10,67 +10,67 @@
         },
         @{
             ModuleName      = 'Microsoft.Graph.Applications'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Authentication'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Beta.DeviceManagement'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Beta.Devices.CorporateManagement'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Beta.DeviceManagement.Administration'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Beta.DeviceManagement.Enrollment'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Beta.Identity.DirectoryManagement'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Beta.Identity.Governance'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Beta.Identity.SignIns'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Beta.Teams'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.DeviceManagement.Administration'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Beta.DirectoryObjects'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Groups'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Planner'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Users'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.Graph.Users.Actions'
-            RequiredVersion = '2.0.0'
+            RequiredVersion = '2.1.0'
         },
         @{
             ModuleName      = 'Microsoft.PowerApps.Administration.PowerShell'

--- a/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
+++ b/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
@@ -2761,6 +2761,12 @@ function Update-M365DSCDependencies
         $ValidateOnly
     )
 
+
+    $Global:MaximumFunctionCount = 16000
+
+    Write-Verbose  "Current value for Maxmimum Function Count: §($global:MaximumFunctionCount)"
+
+
     $InformationPreference = 'Continue'
     $currentPath = Join-Path -Path $PSScriptRoot -ChildPath '..\' -Resolve
     $manifest = Import-PowerShellDataFile "$currentPath/Dependencies/Manifest.psd1"

--- a/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
+++ b/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
@@ -2795,7 +2795,8 @@ function Update-M365DSCDependencies
         }
         catch
         {
-            Write-Host "Could not update {$($dependency.ModuleName)}"
+            Write-Host "Could not update or import {$($dependency.ModuleName)}"
+            Write-Host "Error-Mesage: $($_.Exception.Message)"
         }
         $i++
     }

--- a/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
+++ b/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
@@ -1157,6 +1157,8 @@ function Export-M365DSCConfiguration
 
     $Global:MaximumFunctionCount = 16000
 
+    Write-Verbose  "Current value for Maxmimum Function Count: §($global:MaximumFunctionCount)"
+
     # Define the exported resource instances' names Global variable
     $Global:M365DSCExportedResourceInstancesNames = @()
 
@@ -1660,7 +1662,11 @@ function New-M365DSCConnection
         [System.Boolean]
         $SkipModuleReload = $false
     )
+
     $Global:MaximumFunctionCount = 16000
+
+    Write-Verbose  "Current value for Maxmimum Function Count: §($global:MaximumFunctionCount)"
+
     if ($Workload -eq 'MicrosoftTeams')
     {
         try

--- a/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
+++ b/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
@@ -1665,8 +1665,6 @@ function New-M365DSCConnection
 
     $Global:MaximumFunctionCount = 16000
 
-    Write-Verbose  "Current value for Maxmimum Function Count: §($global:MaximumFunctionCount)"
-
     if ($Workload -eq 'MicrosoftTeams')
     {
         try
@@ -2763,9 +2761,6 @@ function Update-M365DSCDependencies
 
 
     $Global:MaximumFunctionCount = 16000
-
-    Write-Verbose  "Current value for Maxmimum Function Count: §($global:MaximumFunctionCount)"
-
 
     $InformationPreference = 'Continue'
     $currentPath = Join-Path -Path $PSScriptRoot -ChildPath '..\' -Resolve

--- a/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
+++ b/Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1
@@ -1157,8 +1157,6 @@ function Export-M365DSCConfiguration
 
     $Global:MaximumFunctionCount = 16000
 
-    Write-Verbose  "Current value for Maxmimum Function Count: §($global:MaximumFunctionCount)"
-
     # Define the exported resource instances' names Global variable
     $Global:M365DSCExportedResourceInstancesNames = @()
 
@@ -2759,8 +2757,6 @@ function Update-M365DSCDependencies
         $ValidateOnly
     )
 
-
-    $Global:MaximumFunctionCount = 16000
 
     $InformationPreference = 'Continue'
     $currentPath = Join-Path -Path $PSScriptRoot -ChildPath '..\' -Resolve

